File: concurrency/channels/fan_in.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func main() {
	// producers
	ch := make(chan string)
	wg := sync.WaitGroup{}
	messageList := [][]string{
		{"hi", "from", "the", "thread"},
		{"more", "messages", "for", "you"},
	}
	for _, list := range messageList {
		wg.Add(1)
		go producer(ch, &wg, list)
	}

	// close channel
	go func() {
		wg.Wait()
		close(ch)
	}() // avoids deadlock because not waiting to receive before closing channel

	// receive
	for received := range ch {
		fmt.Println("Got: ", received)
	}
}

func producer(out chan<- string, wg *sync.WaitGroup, messages []string) {
	go func() {
		defer wg.Done()
		for _, m := range messages {
			out <- m
			time.Sleep(time.Second * 1)
		}
	}()
}

File: concurrency/channels/fan_in.rs
use std::sync::mpsc; // multiple producer single consumer
use std::thread;
use std::time::Duration;

fn main() {
    let (tx, rx) = mpsc::channel(); 

    spawn_sender(tx.clone(), vec![
        String::from("hi"),
        String::from("from"),
        String::from("the"),
        String::from("thread"),
    ]);
    spawn_sender(tx, vec![ // have to have tx last as clone borrows after move o/w
        String::from("more"),
        String::from("messages"),
        String::from("for"),
        String::from("you"),
    ]);

    for received in rx {
        println!("Got: {received}");
    }
    // unlike golang do not need to close a channel because when all references are dropped to a tx when no senders left
    // note the converse is not true so if receivers destructed the senders will not be - they will error instead
}

fn spawn_sender(tx: mpsc::Sender<String>, messages: Vec<String>) {
    thread::spawn(move || {
        for message in messages {
            tx.send(message).unwrap();
            thread::sleep(Duration::from_secs(1));
        }
    });
}